machine.go

```diff
@@ -538,14 +538,18 @@ func (m *Machine) startup(command string, extracontent [][2]string) (int, error)
 	if err != nil {
 		return -1, err
 	}
+	KernelPath := "/boot/vmlinuz-" + kernelRelease
+	if _, err := os.Stat(KernelPath); err != nil {
+		KernelPath = "/boot/vmlinuz-linux"
+	}
```

Comment:
I'm not against it but it seems very haphazard. How can we be sure the modules in the initramfs match the kernel version ?

Reply:
Does Arch not suppot multiple parallel installed kernels or ?

Reply:
So Arch does support parallel kernel installations, but not in the sense "different versions", but more "different packages, that happen to be different versions". The idea of including the version in the filenames and paths is discussed at: https://bugs.archlinux.org/task/16702. Unfortunately the discussion has been ongoing for 10 years.

The two more common packages that you might want to install are `linux` and `linux-lts` (from the `core` repository), and there is also `linux-zen` and `linux-hardened` from the `extra` repository.

If I install them all, here's what I get:

```
# ls -l /lib/modules
total 16
drwxr-xr-x 3 root root 4096 Dec 10 07:31 4.19.88-1-lts
drwxr-xr-x 3 root root 4096 Dec 10 07:31 5.3.15.a-1-hardened
drwxr-xr-x 3 root root 4096 Dec 10 07:33 5.4.2-arch1-1
drwxr-xr-x 3 root root 4096 Dec 10 07:28 5.4.2-zen1-1-zen

# ls -l /boot/
total 162528
-rw-r--r-- 1 root root 26461234 Dec 10 07:33 initramfs-linux-fallback.img
-rw-r--r-- 1 root root 27048174 Dec 10 07:31 initramfs-linux-hardened-fallback.img
-rw-r--r-- 1 root root  8465635 Dec 10 07:31 initramfs-linux-hardened.img
-rw-r--r-- 1 root root  8411613 Dec 10 07:33 initramfs-linux.img
-rw-r--r-- 1 root root 25952625 Dec 10 07:31 initramfs-linux-lts-fallback.img
-rw-r--r-- 1 root root  8422514 Dec 10 07:31 initramfs-linux-lts.img
-rw-r--r-- 1 root root 27433250 Dec 10 07:28 initramfs-linux-zen-fallback.img
-rw-r--r-- 1 root root  8455012 Dec 10 07:28 initramfs-linux-zen.img
-rw-r--r-- 1 root root  6404480 Dec 10 07:33 vmlinuz-linux
-rw-r--r-- 1 root root  6527360 Dec 10 07:31 vmlinuz-linux-hardened
-rw-r--r-- 1 root root  5896064 Dec 10 07:31 vmlinuz-linux-lts
-rw-r--r-- 1 root root  6924672 Dec 10 07:28 vmlinuz-linux-zen
```

At a quick glance, it's impossible to reliably map what's in `/usr/lib/modules` with the kernels in `/boot` in a generic way. Unless we want to make debos "archlinux aware" and start grepping around for keywords like `lts`, `hardened` and so on, but I'd prefer to avoid that.

I think a better solution would be to scan for files `/boot/vmlinux-*`, and get to know more about these files. For example, with `file(1)` we can actually know the kernel version:

```
$ file /boot/vmlinuz-linux*
/boot/vmlinuz-linux:          Linux kernel x86 boot executable bzImage, version 5.4.2-arch1-1 (linux@archlinux) #1 SMP PREEMPT Thu, 05 Dec 2019 12:29:40 +0000, RO-rootFS, swap_dev 0x6, Normal VGA
/boot/vmlinuz-linux-hardened: Linux kernel x86 boot executable bzImage, version 5.3.15.a-1-hardened (linux-hardened@archlinux) #1 SMP PREEMPT Thu, 05 Dec 2019 22:09:18 +0000, RO-rootFS, swap_dev 0x6, Normal VGA
/boot/vmlinuz-linux-lts:      Linux kernel x86 boot executable bzImage, version 4.19.88-1-lts (linux-lts@archlinux) #1 SMP Thu, 05 Dec 2019 12:30:39 +0000, RO-rootFS, swap_dev 0x5, Normal VGA
/boot/vmlinuz-linux-zen:      Linux kernel x86 boot executable bzImage, version 5.4.2-zen1-1-zen (linux-zen@archlinux) #1 ZEN SMP PREEMPT Thu, 05 Dec 2019 12:30:04 +0000, RO-rootFS, swap_dev 0x6, Normal VGA
```

I wonder if we could either use file and parse the output, or directly open the vmlinuz file and look straight in the right spot to know the version.

Reply:
FWIW, ArchLinux's derivative Manjaro is a bit more similar to Debian, in the sense that it provides different packages to install different specific version of the kernel, for example `linux419` for `4.19`, `linux53` for `5.3`. They also provide packages for realtime kernel, eg. `linux419-rt`.

However, once again you can't directly use the kernel version in `/usr/lib/modules` to find the kernel in `/boot`:

```
# ls -l /usr/lib/modules
drwxr-xr-x 4 root root 4096 Dec 10 08:36 4.19.72-rt26-MANJARO
drwxr-xr-x 4 root root 4096 Dec 10 08:31 4.19.85-1-MANJARO
drwxr-xr-x 4 root root 4096 Dec 10 08:31 5.3.12-1-MANJARO

# ls -l /boot
-rw-r--r-- 1 root root 30711913 Dec 10 08:36 initramfs-4.19-rt-x86_64-fallback.img
-rw-r--r-- 1 root root  8161244 Dec 10 08:36 initramfs-4.19-rt-x86_64.img
-rw-r--r-- 1 root root 30646818 Dec 10 08:31 initramfs-4.19-x86_64-fallback.img
-rw-r--r-- 1 root root  8159883 Dec 10 08:31 initramfs-4.19-x86_64.img
-rw-r--r-- 1 root root 31060316 Dec 10 08:32 initramfs-5.3-x86_64-fallback.img
-rw-r--r-- 1 root root  8168645 Dec 10 08:31 initramfs-5.3-x86_64.img
-rw-r--r-- 1 root root       27 Nov 11 19:00 linux-419-rt-x86_64.kver
-rw-r--r-- 1 root root       22 Nov 21 10:36 linux419-x86_64.kver
-rw-r--r-- 1 root root       21 Nov 21 10:55 linux53-x86_64.kver
-rw-r--r-- 1 root root  5924736 Dec 10 08:36 vmlinuz-4.19-rt-x86_64
-rw-r--r-- 1 root root  5957504 Dec 10 08:31 vmlinuz-4.19-x86_64
-rw-r--r-- 1 root root  6343040 Dec 10 08:31 vmlinuz-5.3-x86_64
```

Once again, using `file` (or `strings`) to find the kernel version would work.

@sjoerd Is it acceptable? `file`, `string` or something else? Or do you think of a better solution?

Reply:
Wonder if there is a golang library that can help here (to get the string out) rather then hacking it via shell commands; But yeah doing that would be a good way going forward

Reply:
Couldn't find a golang library, so I wrote it myself. It's actually trivial to open a file and find a string.